src/fundus/scraping/common_crawl/pipeline.py

```diff
@@ -0,0 +1,274 @@
+                through all articles with at least title, body, and publishing_date set.
+            url_filter (Optional[URLFilter]): A callable object satisfying the URLFilter protocol to skip
+                URLs before download. This filter applies on both requested and responded URL. Defaults to None.
+            only_unique (bool): If set to True, articles yielded will be unique on the responded URL.
+                Always returns the first encountered article. Defaults to True.
+
+        Returns:
+            Iterator[Article]: An iterator yielding objects of type Article.
+        """
+
+        if max_articles == 0:
+            return
+        elif max_articles is None:
+            max_articles = -1
+
+        def build_extraction_filter() -> Optional[ExtractionFilter]:
+            if isinstance(only_complete, bool):
+                return (
+                    None
+                    if only_complete is False
+                    else lambda extracted: not all(
+                        bool(v) if not isinstance(v, Exception) else False for _, v in extracted.items()
+                    )
+                )
+            else:
+                return only_complete
+
+        warc_paths = self._get_list_of_warc_path(start, end)
+        response_cache: Set[str] = set()
+
+        article_task: Callable[[str], Iterator[Article]] = partial(
+            self._fetch_articles,
+            publishers=self.publishers,
+            error_handling=error_handling,
+            extraction_filter=build_extraction_filter(),
+            url_filter=url_filter,
+        )
+
+        if self.processes == -1:
+            article_iter = self._single_crawl(warc_paths, article_task)
+        else:
+            article_iter = self._parallel_crawl(warc_paths, article_task)
+
+        for article_idx, article in enumerate(article_iter, start=1):
+            if not only_unique or article.html.responded_url not in response_cache:
+                response_cache.add(article.html.responded_url)
+                yield article
+            if article_idx == max_articles:
+                break
+
+
+class _PoolResult(Iterable[_T]):
+    def __init__(self, result: MapResult[Any], queue: Queue[_T]):
+        """Utility class to iterate a pool queue.
+
+        Exhaust the queue given with <queue>. If <queue> raises Empty and the pool finished,
+        raise StopIteration, otherwise continue to wait for the next result from <queue>.
+
+        Args:
+            result (MapResult[Any]): A MapResult returned by a Pool.map/.map_async call to use as a handle.
+            queue (Queue[_T): The queue to exhaust.
+        """
+        self._handle = result
+        self._queue = queue
+
+    def __next__(self) -> _T:
+        while True:
+            try:
+                return self._queue.get(timeout=0.1)
+            except Empty:
+                try:
+                    self._handle.get(timeout=0.1)
```

Comment:
Good idea. That's the same as how the pool handles this issue internally. The only thing here is that we have to count the sentinels and somehow i don't like that.

Reply:
Why do we need to count the sentinels? Since we only use a single queue, there would only be a single sentinel as the last retrieved element, wouldn't it?